src/Synthesiser.cpp

```diff
@@ -2213,33 +2213,41 @@ void Synthesiser::generateCode(std::ostream& os, const std::string& id, bool& wi
             }
             os << "}\n";
         }
+    }
 
+    if (!prog.getSubroutines().empty()) {
         // generate subroutine adapter
         os << "void executeSubroutine(std::string name, const std::vector<RamDomain>& args, "
               "std::vector<RamDomain>& ret) override {\n";
-
         // subroutine number
         size_t subroutineNum = 0;
         for (auto& sub : prog.getSubroutines()) {
             os << "if (name == \"" << sub.first << "\") {\n"
-               << "subproof_" << subroutineNum
-               << "(args, ret);\n"  // subproof_i to deal with special characters in relation names
+               << "subroutine_" << subroutineNum
+               << "(args, ret);\n"  // subroutine_<i> to deal with special characters in relation names
+               << "return;"
                << "}\n";
             subroutineNum++;
         }
+        os << "fatal(\"unknown subroutine\");\n";
         os << "}\n";  // end of executeSubroutine
 
         // generate method for each subroutine
         subroutineNum = 0;
         for (auto& sub : prog.getSubroutines()) {
             // method header
             os << "void "
-               << "subproof_" << subroutineNum
+               << "subroutine_" << subroutineNum
                << "(const std::vector<RamDomain>& args, "
                   "std::vector<RamDomain>& ret) {\n";
 
             // a lock is needed when filling the subroutine return vectors
```

Comment:
If the lock is removed, the comment saying the lock is needed is hopefully incorrect and can also be removed.

Reply:
We added the lock for provenance - I was confused to find the lock here. 
Perhaps we can remove it - I need to think it through.

Reply:
Ah, I actually misread the code. I thought that you had removed the lock entirely, but it's actually just moved inside an if statement.